Converge with openspending conductor API

subhankarb opened the pull request:
There were total 3 issues to make dpr-api converge with open-spending API.
- #248 
- #250 
- #251 

Making pull request for three separately was a problem as all three issues were closely related to each other . That is why raising the pull request for all three issues at a time.
Below are the commits for each issues:
- #248 is resolved by https://github.com/subhankarb/dpr-api/commit/136899e43a8d7265100388083c9a2cc55e44344d
- #250 is resolved by https://github.com/subhankarb/dpr-api/commit/f356a6b6c7f5067734f3d52a8ecc2c007f2b7eb6
- #251 is resolved by https://github.com/subhankarb/dpr-api/commit/9ae8201c3dbb92865d0101dc547156c53635ac45

rufuspollock commented:
Commit messages:

First commit message is:

```
[l] Added API /api/bitstore/authorize_upl -refs #248
```

As i understand the commit this is simply false! You are actually adding `/api/datastore/authorise` (?).

For a large commit there is a minimal summary.

rufuspollock commented on app/__init__.py:
```diff
@@ -17,7 +17,7 @@
 from .database import db
 from app.utils import get_s3_cdn_prefix
-from app.auth.controllers import auth_blueprint
+from app.auth.controllers import auth_blueprint, datastore_blueprint
```
Why is it called `datastore_blueprint`. As discussed, this is really about the `bitstore`. We are only using `datastore` in the API to maintain compatibility with OS as it is right now!

rufuspollock commented on tests/auth/test_controller.py:
```diff
@@ -267,4 +269,123 @@ def tearDown(self):
+            db.engine.dispose()
+
+
+class AuthorizeUploadTestCase(unittest.TestCase):
```
I'd assumed this change was more a refactor. If we are adding this test case where are the tests we are deleting?

subhankarb replied:
@rufuspollock removed `GetS3SignedUrlTestCase` in commit 7cb0748b9f099fa0a87b3231cfffb66643b31f62

rufuspollock commented on app/auth/annotations.py:
```diff
@@ -33,23 +29,11 @@ def is_allowed(action):
-                return handle_error("INVALID_ENTITY", "{e} is not a valid one".format(e=entity_str), 401)
-
-            status = ia(user_id, instance, action)
+            status = get_status(action, kwargs['publisher'], kwargs['package'], user_id)
```
Why the refactor of this into a separate file. A note on this in the commit message would have helped enormously.

Also `get_status` is a very undescriptive function name. As I understand it get_status is really "check_is_authorized" (??)

rufuspollock commented on app/auth/controllers.py:
```diff
@@ -222,3 +224,63 @@ def get_s3_signed_url():
         return handle_error('GENERIC_ERROR', e.message, 500)
+
+
+@datastore_blueprint.route('/authorize', methods=['POST'])
```
Where is the route we are deleting and replacing with this route (or redirecting to this route if we are maintaining backwards compatibility)?

rufuspollock commented on app/auth/models.py:
```diff
@@ -152,3 +153,42 @@ def create_user(self, email, full_name, user_name, password):
             return None
+
+
+class FileData(object):
```
Why introduce this new model object?

Did any functionality like this exist before? If so, where and how has that been refactored into this code?

Again a note in the commit message about this would help explain how these changes fit together.

subhankarb replied:
@rufuspollock This was not implemented before. Previously `finalize` API had only one method call from `Bitstore` and get one URL and return back. Right now we have to build a big json to return back to cli. 
That is why i separated the code from controller, so that it can be testable.

rufuspollock commented on app/auth/models.py:
```diff
@@ -152,3 +153,42 @@ def create_user(self, email, full_name, user_name, password):
+                                            **kwargs)
+        return post
+
+    def get_response(self):
```
Why is a model object got stuff about making a response for the API (I assume this is a response for the API??). That could indicate a poor separation of concerns.

rufuspollock commented on tests/package/test_models.py:
```diff
@@ -444,6 +444,12 @@ def test_should_update_data_package_if_preexists(self):
                    Package.version == 'tag_one').one()
         self.assertEqual(latest_data.readme, tagged_data.readme)
 
+    def test_should_return_true_if_package_already_exists(self):
```
I think this test case could be better named -- *what* should return true if package exists ...?

Given it is part of a general `TestPackage` test case, i suggest calling it simply:

```
test_is_package_exists
```

rufuspollock commented on app/package/controllers.py:
```diff
@@ -330,70 +331,66 @@ def purge_data_package(publisher, package):
-                         methods=["POST"])
+@package_blueprint.route("/upload", methods=["POST"])
 @requires_auth
-@is_allowed('Package::Create')
```
I think this should still require authorization. Otherwise anyone can call this and reimport data. Amongst everything else it opens us to potential DDOS since the operation is expensive potentially.

rufuspollock replied:
Ah ok, i now see that below you actually have moved authorization in method (something i agree with and recommended you would want to do back in December!)

subhankarb replied:
@rufuspollock i never thought the url pattern would change.

rufuspollock commented on app/package/controllers.py:
```diff
@@ -330,70 +331,66 @@ def purge_data_package(publisher, package):
 @requires_auth
-@is_allowed('Package::Create')
-def finalize_metadata(publisher, package):
+def finalize_metadata():
```
I'd have renamed the method to something more meaningful. It is not finalizing metadata so much as completing the "publish process" so maybe:

```
finalize_publish
```

rufuspollock commented on app/package/controllers.py:
```diff
@@ -330,70 +331,66 @@ def purge_data_package(publisher, package):
-def finalize_metadata(publisher, package):
+def finalize_metadata():
     """
     DPR metadata finalize operation.
```
Don't use the acronym DPR in the codebase if you can.

The first line here is also a pointless piece of documentation as it tells the reader nothing they don't already know from the method name. I would just delete this and keep the second line "This API is ..." which **is** useful.

subhankarb replied:
@rufuspollock The first line is more of a heading of the swagger document.

rufuspollock commented on app/package/controllers.py:
```diff
@@ -330,70 +331,66 @@ def purge_data_package(publisher, package):
+def finalize_metadata():
     """
     DPR metadata finalize operation.
     This API is responsible for getting data from S3 and push it to RDS.
```
Not just getting data but getting the Data Package (datapackage.json, README) and importing it into our MetaStore

(and later on we may import the data into a DataStore!)

rufuspollock commented on app/utils/auth_helper.py:
```diff
@@ -14,24 +14,15 @@
-    auth = req.headers.get('Authorization', None)
+    token = req.headers.get('Authorization', None)
+    if token is None:
+        token = req.headers.get('Auth-Token', None)
```
I don't think we need to support multiple options do we? If we are switching to `Auth-Token` header let's just go with that.

rufuspollock replied:
Actually, scratch that. Authorization is the standard one so let's stick with that.

subhankarb commented:
We want to converge API of both dpr-api and os-conductor. The epic frictionlessdata/dpr-api#249 addressed all the changes we wanted to make so that dpmpy works with both the APIs.

In epic frictionlessdata/dpr-api#249 we mentioned three issues that is related to dpr-api.

- #248 Create /api/bitstore/authorize endpoint. The Commit 136899e43a8d7265100388083c9a2cc55e44344d contains the changes.
  - Previously our url pattern was `{api}/{publisher}/{package}` so we always get publisher and package information. That is why writing annotation was easy. But with `{api}/bitstore/authorize` we need to get publisher and package information from request body. So i am separating the is_authorize logic from annotation and put it in separate method, so that we can call it from controller.
  - Added model `FileData`. In older implementation we had only one s3 url call. But now we have list of files and have to generate a structured json response. That is why separating json building logic from controller is better. It is more testable.
  - Added a method to check datapackage already exists for validation.
  - Added related tests.

rufuspollock commented:
@subhankarb the summary is **much** better -- well done.

Only thing missing was to have had a classic commit summary first line e.g.

```
[refactor][l]: Refactor to converge our API with openspending conductor API.

...
```
